Latest gtm (#778)
Amplitude-Developer-Docs/amplitude-dev-center: docs/data/sources/google-tag-manager-client.md

![Web Container Setup](../../assets/images/gtm/gtm-web-container-setup.png) 

### Add Template

Create a new tag template by searching the gallery. Choose Amplitude Analytics Browser SDK and click `add` button.
![Amplitude Analytics Browser SDK ](../../assets/images/gtm/gtm-amplitude-analytics-browser-sdk.png) 

### Create Tags

Create tags for your amplitude browser SDK tracking. Click the `New` button to create your new tags.

#### Tag Configuration

Amplitude provides a list of tag types. You might also create a Custom HTML tag to do customization, like cross domain tracking or others.  For successful tracking in Amplitude, ensure that the Amplitude tag loads before the Custom HTML tag, which is being used to call Amplitude. The sequence in which the tags load can significantly influence the outcome. [Time sequence](https://support.google.com/tagmanager/answer/6238868) might be helpful in this case. Please check [this section](./#access-amplitude-instance) for accessing amplitude instance in the Custom HTML tag. Note: We are not recomended to install amplitude through Cusom HTML tag.
 
#### API Key

> Copy your amplitude API Key in the API Key field. For EU recidency, please make your API Key is under analytics.eu.amplitude.com. See [this page](https://tagmanager.google.com/?authuser=0#/container/accounts/6053924196/containers/91603214/workspaces/62/tags:~:text=project%20settings.%20See-,this%20page,-for%20more%20information) for more details. 

Reviewer: For accessibility I think "see" is not recommended terminology. I have been using "Learn more" or "More details" instead.

```suggestion
Copy your amplitude API Key in the API Key field. For EU recidency, please make your API Key is under analytics.eu.amplitude.com. Learn more [here](https://tagmanager.google.com/?authuser=0#/container/accounts/6053924196/containers/91603214/workspaces/62/tags:~:text=project%20settings.%20See-,this%20page,-for%20more%20information). 
```

#### Instance Name

If you plan on running multiple Amplitude instances, each with distinct API keys or initialization options, you can assign an `Instance Name` to each one. It's crucial to configure the `Instance Name` in the tag settings for every Amplitude tag that should be associated with a specific instance. 

!!!note: 
    This approach can also prevent missing events if different versions of the Amplitude SDK coexist in your system.

#### Access amplitude instance 

If you have a Custome HTML tag, you might need to acccess `amplitude` instance in your script tag. 

> ##### With defualt Instance Name 

Reviewer: typo

```js
// For amplitude-js-gtm@3.1.4 and above
amplitudeGTM.getDeviceId();

// others

> amplitude.getDeviceId();

Author: It's not auto-upgraded. You need to click a refresh button. 
I will add the recommendation not only for the instance name but make it in general.

```

##### With customized Instance Name

```js
// For amplitude-js-gtm@3.1.4 and above
amplitudeGTM._iq["yourInstanceName"].getDeviceId();

// Others
amplitude._iq["yourInstanceName"].getDeviceId();
```

### Select a Tag Types

A tag type allow you to specify what kind of action or event should be tracked in your application. The following tag types are supported in Amplitude GTM template.

### init

!!!note
    `init` operates as a separate tag type. It's essential to create a tag for the init tag type. Although we provide defer initialization, events won't be sent to amplitude until the `init` tag has been activated.

#### EU Data Residency

> For EU data residency, you must set up your project inside Amplitude EU and use the API key from Amplitude EU. You can configure the server zone by checking the checkbox **EU Data Residency** under **Tag Configuration** -> **Initialization** of the `init` tag. The initialization section only shows up when tag type is set to init. [More details](../../data/sdks/typescript-browser/#eu-data-residency).

Reviewer: ```suggestion
For EU data residency, you must set up your project inside Amplitude EU and use the API key from Amplitude EU. You can configure the server zone by checking the checkbox **EU Data Residency** under **Tag Configuration** -> **Initialization** of the `init` tag. The initialization section only shows up when tag type is set to `init`. [More details](../../data/sdks/typescript-browser/#eu-data-residency).
```

    |`cookieSecure`| `boolean`. If restrict access to cookies or not. A cookie with the Secure attribute is only sent to the server with an encrypted request over the HTTPS protocol. | `false` |
    |`disableCookies`| `boolean`. If disable cookies or not. If cookies is disable, using LocalStorage or MemoryStorage. | The cookies is enable by default. |
    |`domain`| `string`. Set the top level domain. | `null` |
    |`partnerId`| `string`. The partner Id value. Amplitude requires the customer who built an event ingestion integration to add the partner identifier to `partner_id`. | `null` |
    |`sessionTimeout`| `number`. How long one session expire. | `30` minutes. |
    |`userId`| `number`. ID for the user. Must have a minimum length of 5 characters unless overridden with the `min_user_length` option. | `undefined` |
    |`trackingOptions`| `TrackingOptions`. Please check the `Optional tracking` section for more tracking options configuration. | Enable all tracking options by default. |
    |`transport`| `TransportType.XHR` or `TransportType.SendBeacon` or `TransportType.Fetch`. Set the transport type. | `TransportType.Fetch` |

> - `Select a {{GTM variable}} from the list`. It's necessary to return an object containing the key-value pairs you wish to use for instance configuration. Ensure that the keys are part of the available configurations.

Reviewer 2: 👍 Didn't know how to use `{{GTM variable}}` to configure before. 

![image](https://github.com/Amplitude-Developer-Docs/amplitude-dev-center/assets/31029607/71ba0d92-60ca-4d89-8206-97591597b7eb)

Curious if `{{GTM variable}}`s in the dropdown are pre-defined? Can they be used out-of-box?

Author: No. Actually, I try to summarize the options except `Use default values` and `Set configuration manually`. The others are all GTM variable, isn't it?

Author: I removed the `{{}}`. Hope that's clearer.

#### setDeviceId

If you want to reset the userId and deviceId after logout, please check `reset` tag type instead.

| Name | Description | Default Value |
| --- | --- | --- |
| `Device ID`| `string`. Set the deviceId for the current user. Amplitude will assign an unique identifier for the deviceId by default. For the cross domain use case, Amplitue will auto capture the `deviceId` from URL parameter and assign the value to deviceId. Otherwise, this is not recommended unless you know what you are doing. [More Details](https://www.docs.developers.amplitude.com/data/sdks/typescript-browser/#custom-device-id) | `UUID` |

#### setSessionId

Session logic is auto handled by Amplitude. You might need to set the sessionId to `-1` if you want to out of session control. 
Please make sure the value is in milliseconds since epoch (Unix Timestamp) or `-1`.

[More details](../../data/sdks/typescript-browser/#custom-session-id)

#### reset 

`reset` will be commonly used when a user has been logged out. It includes 2 operations which are `setUserId(undefined)` and `setDeviceId(UUID())`.

[More details](/sdks/typescript-browser/#reset-when-user-logs-out).

#### setOptOut

> Check the `Opt current user out of tracking` checkbox to opt user out of tracking. 

Reviewer 2: How about adding some cookie behavior description here or at init tag type

Cookies are created on initialization. So even if a user opts out, cookies are still created. If you want to don't want to track cookies, for example, before user consents, you should wait to init

Author: I added this in init section.

[More details](sdks/typescript-browser/#opt-users-out-of-tracking).

### Define your trigger - Triggering

All tags fire based on events. Anytime Google Tag Manager registers an event, event triggers are evaluated and tags are fired accordingly. Please check [here](https://support.google.com/tagmanager/topic/7679108?sjid=4311393792871502449-NA) for avaialbe triggers in GTM. 

!!!note
    Tipically `init` tag is the first thing you need to fire on the page. It's common to fire the `init` tag using `All Pages` or `Initialization - All Pages` triggers. You can also defer the `init` tag until you receive a signal and fire it using customized triggers, such as a consent grant. But notice that all other tags will wait for the `init` tag to fire before they can be sent to Amplitude. 

> ## Others

Reviewer: ```suggestion
## Video Tutorial
```

description: Collect data with ease using Amplitude Analytics Browser SDK GTM template - the official client-side Google Tag Manager template for seamless data collection.
---

> This is the client-side Google Tag Manager Template for Amplitude Analytics. The tag uses the [Amplitude Marketing Analytics SDK](../../sdks/marketing-analytics-browser/) for data collection. However, due to inherent limitations of GTM, certain features, such as plugins, are not well supported in this GTM template. You are still able to add plugins using the Custom HTML tag, but because of how SDK get loaded GTM, this could lead missing data.

Reviewer: This might be better in an admonishment - note/warning

> However, due to inherent limitations of GTM, certain features, such as plugins, are not well supported in this GTM template. You are still able to add plugins using the Custom HTML tag, but because of how SDK get loaded GTM, this could lead missing data.

#### reset 

`reset` will be commonly used when a user has been logged out. It includes 2 operations which are `setUserId(undefined)` and `setDeviceId(UUID())`. [More details](../../sdks/typescript-browser/#reset-when-user-logs-out).

#### setOptOut

Check the `Opt current user out of tracking` checkbox to opt user out of tracking. [More details](../../sdks/typescript-browser/#opt-users-out-of-tracking).

### Define your trigger - Triggering

All tags fire based on events. Anytime Google Tag Manager registers an event, event triggers are evaluated and tags are fired accordingly. Please check [here](https://support.google.com/tagmanager/topic/7679108?sjid=4311393792871502449-NA) for avaialbe triggers in GTM. 

!!!note
    Tipically `init` tag is the first thing you need to fire on the page. It's common to fire the `init` tag using `All Pages` or `Initialization - All Pages` triggers. You can also defer the `init` tag until you receive a signal and fire it using customized triggers, such as a consent grant. But notice that all other tags will wait for the `init` tag to fire before they can be sent to Amplitude. 

## Common Issues

> ### Is Amplitude GTM template support React Native (Android or iOS) app?

Reviewer: ```suggestion
### Does Amplitude GTM template support React Native (Android or iOS) apps?
```

> No. The Amplitude Analytics Browser SDK only suport the web conatiner, designed for desktop and mobile browsers. Amplitude does not provide support for GTM templates for mobile apps at this time. To request this feature, please submit your request. Please submit a feature request at [here](https://help.amplitude.com/hc/en-us/requests/new).

Reviewer: ```suggestion
No. The Amplitude Analytics Browser SDK only support the web container, designed for desktop and mobile browsers. Amplitude does not provide support for GTM templates for mobile apps at this time. To request this feature, please submit your request. Please submit a feature request at [here](https://help.amplitude.com/hc/en-us/requests/new).
```

### Can we add some logic to Amplitude template?

> Yes, but not recommended. Modified Community Gallery Templates will no longer receive updates provided by the developer but you can always restore the template to start receiving notifications again.

Reviewer: ```suggestion
Yes, but it's not recommended. Modified Community Gallery Templates will no longer receive updates provided by the developer but you can always restore the template to start receiving notifications again.
```
